rows.go

```diff
@@ -85,3 +161,40 @@ func (rows *doltRows) Next(dest []driver.Value) error {
 
 	return nil
 }
+
+// peekableRowIter wrap another gms.RowIter and allows the caller to peek at results, without disturbing the order
+// that results are returned from the Next() method.
+type peekableRowIter struct {
+	iter  gms.RowIter
+	peeks []gms.Row
+}
+
+var _ gms.RowIter = (*peekableRowIter)(nil)
+
+// Peek returns the next row from this row iterator, without causing that row to be skipped from future calls
+// to Next(). There is no limit on how many rows can be peeked.
+func (p *peekableRowIter) Peek(ctx *gms.Context) (gms.Row, error) {
+	next, err := p.iter.Next(ctx)
+	if err != nil {
+		return nil, err
```

Comment:
Does this mean Peek returns a NoRow error or similar if the result set is empty?

Reply:
Yup, `Peek()` will return `io.EOF` if the result set is empty. It doesn't need to store the error, since calling `Next()` again will return `io.EOF` again (or any other error it encountered).